src/aggregate/aggregate_request.c

```diff
@@ -1210,6 +1210,53 @@ int buildOutputPipeline(AREQ *req, QueryError *status) {
   return REDISMODULE_ERR;
 }
 
+// Add rp_to_place to an existing pipeline as the upstream result proceesor of rp.
+static void PushUpStream(ResultProcessor *rp_to_place, ResultProcessor *rp) {
+  rp_to_place->upstream = rp->upstream;
+  rp->upstream = rp_to_place;
+  rp_to_place->parent = rp->parent;
+}
+
+// Add Buffer-Locker and Unlocker result processors to the pipeline.
+// The Buffer-Locker rp is added as the upstream of the first result processor that might
+// access Redis keyspace.
+// The Unlocker is places so that its upstream rp will be the last to access Redis keyspace.
+static int SafeRedisKeyspaceAccessPipeline(AREQ *req, ResultProcessor *first_to_access_redis,
+                                    ResultProcessor *last_to_access_redis, QueryError *status) {
+  
+  // TODO: multithreaded: Add better estimation to the buffer initial size
+  ResultProcessor *rpBufferAndLocker = RPBufferAndLocker_New(1024);
+
+  // Place buffer and locker as the upstream of the first_to_access_redis result processor.
+  PushUpStream(rpBufferAndLocker, first_to_access_redis);
+
+  // Find where to place unlocker
+  ResultProcessor *rpUnlocker = RPUnlocker_New((RPBufferAndLocker*)rpBufferAndLocker);
+
+  // Start from the end processor and iterate beackward until the next rp
+  // is the last to access redis and push the unlocker between them.
+
+  ResultProcessor dummy_rp = {.upstream = req->qiter.endProc};
+  ResultProcessor *curr_rp = &dummy_rp;
+
+  while (last_to_access_redis != curr_rp->upstream && curr_rp != req->qiter.rootProc) {
+    curr_rp = curr_rp->upstream;
+  }
+  // If we didn't find where to push the unlocker something went wrong...
+  if (curr_rp == req->qiter.rootProc) {
+    QueryError_SetErrorFmt(status, QUERY_ECONSTRUCT_PIPELINE, "Can't find unlocker position in the query stream");
+    return REDISMODULE_ERR;
+  }
+  // Handle special case where endProc is the last.

```

Comment:
in which case this will happen?

Reply:
In case we forget to set the `last_access` rp. The point is to avoid an infinite loop.
I changed it to a more strict condition - check that we didn't pass the buffer-locker (no sense placing the unlocker right after it... )